File: 程式設計：資料結構答案檔/C#/305.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Employee
{
    public string EId { get; set; }
    public string EName { get; set; }
    public int Salary { get; set; }
    public string DId { get; set; }
}

class Department
{
    public string DId { get; set; }
    public string DName { get; set; }
    public string MId { get; set; }
}

class Project
{
    public string PId { get; set; }
    public string PName { get; set; }
    public string Location { get; set; }
}

class HourLog
{
    public string EId { get; set; }
    public string PId { get; set; }
    public int PHours { get; set; }
}

class ANS
{
    public string PName { get; set; }
    public int PHours { get; set; }
}

class Program
{
    static void Main()
    {
        var numInputs = Console.ReadLine().Split(' ').Select(int.Parse).ToArray();
        var employees = new List<Employee>();
        var departments = new List<Department>();
        var projects = new List<Project>();
        var hourLogs = new List<HourLog>();

        for (int i = 0; i < numInputs[0]; i++)
        {
            var inputs = Console.ReadLine().Split(' ');
            employees.Add(new Employee { EId = inputs[0], EName = inputs[1], Salary = int.Parse(inputs[2]), DId = inputs[3] });
        }

        for (int i = 0; i < numInputs[1]; i++)
        {
            var inputs = Console.ReadLine().Split(' ');
            departments.Add(new Department { DId = inputs[0], DName = inputs[1], MId = inputs[2] });
        }

        for (int i = 0; i < numInputs[2]; i++)
        {
            var inputs = Console.ReadLine().Split(' ');
            projects.Add(new Project { PId = inputs[0], PName = inputs[1], Location = inputs[2] });
        }

        for (int i = 0; i < numInputs[3]; i++)
        {
            var inputs = Console.ReadLine().Split(' ');
            hourLogs.Add(new HourLog { EId = inputs[0], PId = inputs[1], PHours = int.Parse(inputs[2]) });
        }

        string sName = Console.ReadLine();
        var employee = employees.FirstOrDefault(e => e.EName == sName);

        var ansList = new List<ANS>();
        if (employee != null)
        {
            foreach (var log in hourLogs.Where(h => h.EId == employee.EId))
            {
                var project = projects.FirstOrDefault(p => p.PId == log.PId);
                if (project != null)
                {
                    ansList.Add(new ANS { PName = project.PName, PHours = log.PHours });
                }
            }
        }

        int totalHours = ansList.Sum(a => a.PHours);
        ansList.OrderBy(a => a.PName).ToList().ForEach(a => Console.WriteLine($"{a.PName} {a.PHours}"));
        Console.WriteLine(totalHours);
    }
}

File: 程式設計：資料結構答案檔/Python/305.py
class Employee:
    def __init__(self, e_id, e_name, salary, d_id):
        self.e_id = e_id
        self.e_name = e_name
        self.salary = salary
        self.d_id = d_id

class Department:
    def __init__(self, d_id, d_name, m_id):
        self.d_id = d_id
        self.d_name = d_name
        self.m_id = m_id

class Project:
    def __init__(self, p_id, p_name, location):
        self.p_id = p_id
        self.p_name = p_name
        self.location = location

class HourLog:
    def __init__(self, e_id, p_id, p_hours):
        self.e_id = e_id
        self.p_id = p_id
        self.p_hours = p_hours

def getData(employees: list, departments: list, projects: list, hour_logs):
    E, D, P, H = map(int, input().split())
    for _ in range(E):
        e_id, e_name, salary, d_id = input().split()
        employees.append(Employee(e_id, e_name, int(salary), d_id))
    for _ in range(D):
        d_id, d_name, m_id = input().split()
        departments.append(Department(d_id, d_name, m_id))
    for _ in range(P):
        p_id, p_name, p_location = input().split()
        projects.append(Project(p_id, p_name, p_location))
    for _ in range(H):
        e_id, p_id, p_hours = input().split()
        hour_logs.append(HourLog(e_id, p_id, int(p_hours)))
    search_name = input()
    return search_name
        
def check(employees: list, departments: list, projects: list, hour_logs, e_name):
    hours = 0
    ans = []
    employee = [employee for employee in employees if employee.e_name == e_name]
    if employee ==None: return
    for hour_log in hour_logs:        
        if hour_log.e_id == employee[0].e_id:
            e_projects = [project for project in projects if project.p_id == hour_log.p_id]
            if e_projects:
                for project in e_projects:
                    hours += hour_log.p_hours
                    ans.append([project.p_name, hour_log.p_hours])
    f_ans = sorted(ans, key = lambda x: x[0])
    for a in f_ans:
        print(a[0], a[1])
    print(hours)                

def main():
    employees = []
    departments = []
    projects = []
    hour_logs = []   
    e_name = getData(employees, departments, projects, hour_logs)
    check(employees, departments, projects, hour_logs, e_name)    

main()